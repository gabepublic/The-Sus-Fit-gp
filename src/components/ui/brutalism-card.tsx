"use client"

import { cn } from "@/lib/utils"
import { Upload } from "lucide-react"
import { useState, useId } from "react"

interface BrutalismCardProps {
  className?: string
  title?: string
  children?: React.ReactNode
  buttonPosition?: "left" | "right"
  backgroundImage?: string
  shadowRotation?: string
  onImageUpload?: (imageUrl: string) => void
}

export function BrutalismCard({
  className,
  title = "Upload Your Angle",
  children,
  buttonPosition = "left",
  backgroundImage = '/images/zestyVogueColor.jpg',
  shadowRotation = "rotate-0",
  onImageUpload
}: BrutalismCardProps) {
  const [uploadedImage, setUploadedImage] = useState<string | null>(null)
  const [isDragging, setIsDragging] = useState(false)
  const fileInputId = useId()

  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files?.[0]
    if (file) {
      const reader = new FileReader()
      reader.onload = (e) => {
        const imageUrl = e.target?.result as string
        setUploadedImage(imageUrl)
        if (onImageUpload) {
          onImageUpload(imageUrl)
        }
      }
      reader.readAsDataURL(file)
    }
  }

  const handleDragOver = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault()
    setIsDragging(true)
  }

  const handleDragLeave = () => {
    setIsDragging(false)
  }

  const handleDrop = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault()
    setIsDragging(false)

    const file = e.dataTransfer.files?.[0]
    if (file && file.type.startsWith('image/')) {
      const reader = new FileReader()
      reader.onload = (e) => {
        const imageUrl = e.target?.result as string
        setUploadedImage(imageUrl)
        if (onImageUpload) {
          onImageUpload(imageUrl)
        }
      }
      reader.readAsDataURL(file)
    }
  }

  return (
    <div className={cn("relative group p-10", className)}>
      {/* Shadow layers - creating the brutalism offset effect */}
      <div className={cn("absolute top-8 left-[9%] w-[91%] h-[93%] bg-[var(--color-susfit-teal)] rounded-sm transform border-2 border-black", shadowRotation)} />
      {/* Main card */}
      <div
        className={cn(
          "relative bg-[#7BB3D9] border-2 border-black rounded-md p-4 transform transition-transform group-hover:translate-x-1 group-hover:translate-y-1 h-full before:absolute before:inset-0",
          isDragging ? "border-dashed border-4 border-[var(--color-susfit-green)]" : "",
          !uploadedImage ? "before:bg-gradient-to-t before:from-white/30 before:via-black/10 before:to-white/20 before:backdrop-blur-[2px] before:rounded-md before:pointer-events-none" : ""
        )}
        style={{
          backgroundImage: uploadedImage ? `url('${uploadedImage}')` : `url('${backgroundImage}')`,
          backgroundSize: 'cover',
          backgroundPosition: 'center'
        }}
        onDragOver={handleDragOver}
        onDragLeave={handleDragLeave}
        onDrop={handleDrop}
      >
        {/* Hidden file input */}
        <input
          type="file"
          id={fileInputId}
          className="hidden"
          accept="image/*"
          onChange={handleFileChange}
        />

        {/* Upload icon area - only show if no image uploaded */}
        {!uploadedImage && (
          <div className="absolute top-[20%] left-1/2 transform -translate-x-1/2 z-10">
            <div className="text-[var(--color-susfit-green)] text-6xl">
              <svg
                width="120"
                height="120"
                viewBox="0 0 24 24"
                fill="none"
                stroke="currentColor"
                strokeWidth="1"
                strokeLinecap="round"
                strokeLinejoin="round"
                className="transform"
              >
                <path d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4" />
                <polyline points="7,10 12,15 17,10" />
                <line x1="12" y1="15" x2="12" y2="3" />
              </svg>
            </div>
          </div>
        )}

        {/* Children content */}
        {children}

        {/* Title bar button */}
        <div className={cn("absolute bottom-[15%] w-[calc(100%+16px)] transform",
          buttonPosition === "left" ? "-left-10" : "-right-14")}>
          <div className="relative">
            {/* Shadow/offset effect */}
            <div className={cn("absolute w-full h-full bg-[var(--color-susfit-teal)] rounded-md border-2 border-black",
              buttonPosition === "left" ? "-bottom-2 -right-2" : "-bottom-2 -left-2")} />
            {/* Main button */}
            <label
              htmlFor={fileInputId}
              className="relative w-full bg-[var(--color-susfit-pink)] border-2 border-black px-4 py-2 rounded-md font-bold text-black text-lg block text-center cursor-pointer"
            >
              {title}
            </label>
          </div>
        </div>
      </div>
    </div>
  )
}
