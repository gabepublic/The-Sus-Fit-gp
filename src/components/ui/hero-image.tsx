"use client"

import Image from "next/image"
import { cn } from "@/lib/utils"

interface HeroImageProps {
  src: string
  alt: string
  className?: string
  priority?: boolean
}

export function HeroImage({ 
  src, 
  alt, 
  className,
  priority = true 
}: HeroImageProps) {
  return (
    <div className={cn(
      "relative w-full flex items-center justify-center",
      "h-[50vh] min-h-[400px] max-h-[800px]",
      className
    )}>
      <div className="relative w-full h-full max-w-5xl">
        <Image
          src={src}
          alt={alt}
          fill
          priority={priority}
          className="object-contain drop-shadow-2xl"
          sizes="(max-width: 768px) 100vw, (max-width: 1200px) 90vw, 80vw"
        />
      </div>
    </div>
  )
}
