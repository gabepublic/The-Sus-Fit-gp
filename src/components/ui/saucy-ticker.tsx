import React from 'react';
import Marquee from 'react-fast-marquee';

const SaucyTicker: React.FC = () => {
    const phrases = [
        "Posin'",
        "That's my angle",
        "Just for the gram",
        "Actin' brand new",
        "Cappin'",
        "Lookin' Fab",
        "Glow Up Suspiciously",
        "Fake it 'til you make it",
    ];

    return (
        <div className="ticker-container">
            <Marquee
                speed={50}
                gradient={false}
                pauseOnHover={true}
                direction="right"
                delay={0}
                loop={0} // 0 = infinite
            >
                {phrases.map((phrase, index) => (
                    <span key={index} className="ticker-item">
                  {phrase}
                        {/* Add separator between items */}
                        {index < phrases.length - 1 && <span className="separator"> â€¢ </span>}
          </span>
                ))}
            </Marquee>
        </div>
    );
};

export {SaucyTicker};
