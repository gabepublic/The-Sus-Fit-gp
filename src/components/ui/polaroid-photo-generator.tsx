"use client"

import { cn } from "@/lib/utils"
import { useState, useEffect } from "react"
import { Button } from "@/components/ui/button"
import Image from "next/image"

interface PolaroidPhotoGeneratorProps {
    className?: string
    onGenerationStart?: () => void
    onGenerationComplete?: (imageUrl: string) => void
    onClose?: () => void
    onRetry?: () => void
    isGenerating?: boolean
    mockImageUrl?: string
    generatedImage?: string     // base-64 PNG returned from backend
    isLoading?: boolean         // when backend call is in flight
}

export function PolaroidPhotoGenerator({
    className,
    onGenerationStart,
    onGenerationComplete,
    onClose,
    onRetry,
    isGenerating = false,
    mockImageUrl,
    generatedImage,
    isLoading = false
}: PolaroidPhotoGeneratorProps) {
    const [imgLoaded, setImgLoaded] = useState(false)
    const [showButtons, setShowButtons] = useState(false)
    const [progress, setProgress] = useState(0)

    // Debug logging
    console.log('PolaroidPhotoGenerator props:', {
        generatedImage: generatedImage ? `base64 (${generatedImage.length} chars)` : 'null',
        isGenerating,
        isLoading
    })

    // Handle generatedImage changes - show immediately when available
    useEffect(() => {
        if (generatedImage != null) {
            console.log('Generated image available, showing immediately')
            setShowButtons(true)
            setImgLoaded(false) // Will be set to true when image loads
            
            // Call onGenerationComplete immediately without delay
            if (onGenerationComplete) {
                const fullImageUrl = `data:image/png;base64,${generatedImage}`
                onGenerationComplete(fullImageUrl)
            }
        } else {
            setShowButtons(false)
        }
    }, [generatedImage, onGenerationComplete])

    // Handle generation start and progress
    useEffect(() => {
        if (isGenerating && onGenerationStart) {
            onGenerationStart()
            setProgress(0)
            
            // Start progress animation
            const interval = setInterval(() => {
                setProgress(prev => {
                    if (prev >= 100) {
                        clearInterval(interval)
                        return 100
                    }
                    return prev + 2
                })
            }, 60) // Update every 60ms for smooth animation
            
            return () => clearInterval(interval)
        } else {
            setProgress(0)
        }
    }, [isGenerating, onGenerationStart])

    const handleRetry = () => {
        setImgLoaded(false)
        setShowButtons(false)
        if (onRetry) {
            onRetry()
        }
    }

    const handleImageLoad = (e: React.SyntheticEvent<HTMLImageElement>) => {
        const img = e.currentTarget
        console.log('Image loaded successfully')
        console.log('Image dimensions:', {
            naturalWidth: img.naturalWidth,
            naturalHeight: img.naturalHeight,
            clientWidth: img.clientWidth,
            clientHeight: img.clientHeight
        })
        setImgLoaded(true)
    }

    return (
        <div className={cn("relative flex flex-col items-center", className)} data-testid="polaroid">
            {/* Polaroid Container */}
            <div className="relative bg-white p-6 w-[475px] h-[550px] flex flex-col shadow-lg rounded-sm">
                {/* Photo Area */}
                <div className="relative w-full h-[400px] overflow-hidden rounded-sm bg-gray-100">
                    {/* Loading State with Image, Spinner, and Progress Bar (for isGenerating prop) */}
                    {isGenerating && !generatedImage && (
                        <div className="absolute inset-0 flex items-center justify-center">
                            <Image
                                src={mockImageUrl || '/images/demo/WillShalom.jpg'}
                                alt="Generated try-on preview"
                                fill
                                className="object-contain rounded-sm opacity-50"
                                aria-busy="true"
                                unoptimized
                            />
                            <div className="absolute inset-0 flex items-center justify-center bg-black bg-opacity-20">
                                <div className="text-center">
                                    <div className="animate-spin rounded-full h-12 w-12 border-b-2 border-[var(--color-susfit-pink-500)] mx-auto mb-4"></div>
                                    <div className="text-gray-600 font-mono">Generating...</div>
                                    
                                    {/* Progress Bar */}
                                    <div className="mt-4 w-48 h-2 bg-gray-300 rounded-full overflow-hidden">
                                        <div 
                                            className="h-full bg-gradient-to-r from-[var(--color-susfit-pink-500)] to-[var(--color-susfit-teal-500)] transition-all duration-300 ease-out"
                                            style={{ width: `${progress}%` }}
                                            data-testid="progress-bar"
                                        ></div>
                                    </div>
                                </div>
                            </div>
                            <div className="absolute inset-0 animate-pulse bg-gray-200 rounded-sm"></div>
                        </div>
                    )}

                    {/* Loading State with Image and Shimmer (for isLoading prop) */}
                    {isLoading && !generatedImage && (
                        <div className="absolute inset-0 flex items-center justify-center">
                            <Image
                                src={mockImageUrl || '/images/demo/WillShalom.jpg'}
                                alt="Generated try-on preview"
                                fill
                                className="object-contain rounded-sm opacity-50"
                                aria-busy="true"
                                unoptimized
                            />
                            <div className="absolute inset-0 flex items-center justify-center bg-black bg-opacity-20">
                                <div className="text-center">
                                    <div className="animate-pulse rounded-full h-12 w-12 bg-[var(--color-susfit-pink-500)] mx-auto mb-4"></div>
                                    <div className="text-gray-600 font-mono">Loading...</div>
                                </div>
                            </div>
                        </div>
                    )}

                    {/* Generated Image with Fade-in Effect */}
                    {generatedImage && (
                        <div className="absolute inset-0 flex items-center justify-center">
                            <Image
                                src={`data:image/png;base64,${generatedImage}`}
                                alt="Generated try-on preview"
                                fill
                                className={`object-contain rounded-sm transition-opacity duration-300 ${imgLoaded ? 'opacity-100' : 'opacity-0'}`}
                                onLoad={handleImageLoad}
                                onError={(e) => {
                                    console.error('Image failed to load:', e)
                                    console.error('Generated image length:', generatedImage?.length || 0)
                                }}
                                unoptimized
                            />
                        </div>
                    )}

                    {/* Default Image when no other state is active and no mockImageUrl provided */}
                    {!generatedImage && !isGenerating && !isLoading && mockImageUrl === undefined && (
                        <div className="absolute inset-0 flex items-center justify-center">
                            <Image
                                src="/images/demo/WillShalom.jpg"
                                alt="Generated try-on preview"
                                fill
                                className="object-contain rounded-sm"
                                unoptimized
                            />
                        </div>
                    )}

                    {/* Text placeholder when mockImageUrl is explicitly provided */}
                    {!generatedImage && !isGenerating && !isLoading && mockImageUrl !== undefined && (
                        <div className="absolute inset-0 flex items-center justify-center">
                            <div className="text-gray-400 font-mono text-center">
                                Ready to generate
                            </div>
                        </div>
                    )}
                </div>

                {/* Polaroid Bottom Border */}
                <div className="h-[70px] bg-white flex items-center justify-center">
                    <div className="text-lg text-gray-600 font-mono">
                        {isGenerating && !generatedImage && "Processing..."}
                        {isLoading && !generatedImage && "Loading..."}
                        {generatedImage && "Complete!"}
                        {!isGenerating && !isLoading && !generatedImage && "Ready to generate"}
                    </div>
                </div>
            </div>

            {/* Buttons */}
            {showButtons && (
                <div className="mt-4 flex gap-2 justify-center">
                    <Button
                        onClick={handleRetry}
                        className="bg-[var(--color-susfit-pink-500)] hover:bg-[var(--color-susfit-pink-800)] text-black border-2 border-black font-bold text-sm px-3 py-1 transition-all duration-200"
                        aria-label="Retry generation"
                    >
                        Retry
                    </Button>
                    <Button
                        onClick={onClose}
                        className="bg-[var(--color-susfit-teal-500)] hover:bg-[var(--color-susfit-teal-800)] text-white border-2 border-black font-bold text-sm px-3 py-1 transition-all duration-200"
                        aria-label="Close polaroid"
                    >
                        Close
                    </Button>
                </div>
            )}

            {/* Additional retry button */}
            {generatedImage && (
                <div className="mt-4 text-center" aria-live="polite">
                    <button 
                        onClick={onRetry} 
                        className="text-sm underline text-blue-600 hover:text-blue-800 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2 rounded px-2 py-1 transition-colors duration-200"
                        aria-label="Try another outfit"
                    >
                        Try another outfit
                    </button>
                </div>
            )}

            {/* Debug Info */}
            {process.env.NODE_ENV === 'development' && (
                <div className="mt-2 text-xs text-gray-500 font-mono text-center">
                    Generated: {generatedImage ? '✅' : '❌'} | Loading: {imgLoaded ? '✅' : '❌'} | isGenerating: {isGenerating ? 'true' : 'false'}
                </div>
            )}
        </div>
    )
}
