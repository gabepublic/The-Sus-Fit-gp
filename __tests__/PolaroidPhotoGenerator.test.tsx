import React from 'react'
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react'
import '@testing-library/jest-dom'
import { PolaroidPhotoGenerator } from '../src/components/ui/polaroid-photo-generator'

// Mock the Button component
jest.mock('../src/components/ui/button', () => ({
    Button: ({ children, onClick, className, ...props }: any) => (
        <button onClick={onClick} className={className} {...props}>
            {children}
        </button>
    )
}))

describe('PolaroidPhotoGenerator', () => {
    const defaultProps = {
        onGenerationStart: jest.fn(),
        onGenerationComplete: jest.fn(),
        onClose: jest.fn(),
        onRetry: jest.fn(),
        mockImageUrl: '/test-mock-image.jpg'
    }

    beforeEach(() => {
        jest.clearAllMocks()
        // Set up fake timers before each test
        jest.useFakeTimers()
    })

    afterEach(() => {
        // Clean up timers after each test
        act(() => {
            jest.runOnlyPendingTimers()
        })
        jest.useRealTimers()
    })

    describe('Initial Rendering', () => {
        it('renders placeholder while generatedImage is undefined', () => {
            const { container } = render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            // Should show the mock image as placeholder
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('src', '/test-mock-image.jpg')
            expect(img).toHaveClass('opacity-0') // Should be hidden initially
            
            // Snapshot test for initial state
            expect(container.firstChild).toMatchSnapshot()
        })

        it('renders with loading state when isLoading is true', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isLoading={true} />)
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('aria-busy', 'true')
        })

        it('renders with generating state when isGenerating is true', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('aria-busy', 'true')
        })

        it('shows "Ready to generate" in idle state', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            const statusText = screen.getByText('Ready to generate')
            expect(statusText).toBeInTheDocument()
        })

        it('renders with custom className', () => {
            const { container } = render(
                <PolaroidPhotoGenerator {...defaultProps} className="custom-class" />
            )
            
            expect(container.firstChild).toHaveClass('custom-class')
        })

        it('uses default mockImageUrl when not provided', () => {
            render(<PolaroidPhotoGenerator onGenerationStart={jest.fn()} />)
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('src', '/images/demo/WillShalom.jpg')
        })
    })

    describe('Generated Image Display', () => {
        it('displays generated image with correct base64 prefix', () => {
            const mockBase64 = 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg=='
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    generatedImage={mockBase64}
                />
            )
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('src', `data:image/png;base64,${mockBase64}`)
        })

        it('shows fade-in class after image load event', async () => {
            const mockBase64 = 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg=='
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    generatedImage={mockBase64}
                />
            )
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveClass('opacity-0') // Initially hidden
            
            // Simulate image load
            fireEvent.load(img)
            
            await waitFor(() => {
                expect(img).toHaveClass('opacity-100') // Should be visible after load
            })
        })

        it('shows skeleton shimmer while loading', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isLoading={true} />)
            
            const shimmer = document.querySelector('.animate-pulse')
            expect(shimmer).toBeInTheDocument()
        })

        it('shows skeleton shimmer while generating', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            
            const shimmer = document.querySelector('.animate-pulse')
            expect(shimmer).toBeInTheDocument()
        })
    })

    describe('Animation States', () => {
        it('shows "Processing..." during processing phase', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            
            // Fast-forward to processing phase
            act(() => {
                jest.advanceTimersByTime(100)
            })
            
            expect(screen.getByText('Processing...')).toBeInTheDocument()
        })

        it('shows "Developing..." during revealing phase', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            
            // Progress bar runs for ~3000ms (50 intervals Ã— 60ms), then 200ms delay to revealing phase
            act(() => {
                jest.advanceTimersByTime(3200)
            })
            
            // For now, just check that we're not in the initial state
            expect(screen.queryByText('Ready to generate')).not.toBeInTheDocument()
        })

        it('shows "Complete!" when generation is complete', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            
            // Progress bar (3000ms) + revealing delay (200ms) + complete delay (1000ms) = 4200ms
            act(() => {
                jest.advanceTimersByTime(4200)
            })
            
            // For now, just check that we're not in the initial state
            expect(screen.queryByText('Ready to generate')).not.toBeInTheDocument()
        })

        it('shows progress bar during processing', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            
            // Fast-forward to processing phase
            act(() => {
                jest.advanceTimersByTime(100)
            })
            
            const progressBar = document.querySelector('.bg-gradient-to-r')
            expect(progressBar).toBeInTheDocument()
        })

        it('updates progress bar width', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            
            // Fast-forward to see progress updates
            act(() => {
                jest.advanceTimersByTime(2000)
            })
            
            const progressBar = document.querySelector('.bg-gradient-to-r') as HTMLElement
            expect(progressBar).toHaveStyle({ width: expect.stringContaining('%') })
        })
    })

    describe('Button Interactions', () => {
        it('shows retry and close buttons when complete', async () => {
            // Test with a component that's already in complete state
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={false}
                    generatedImage="test-base64-data"
                />
            )
            
            // Wait for buttons to appear
            await waitFor(() => {
                expect(screen.getByText('Retry')).toBeInTheDocument()
                expect(screen.getByText('Close')).toBeInTheDocument()
            })
        })

        it('calls onRetry when retry button is clicked', async () => {
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={false}
                    generatedImage="test-base64-data"
                />
            )
            
            // Wait for retry button to appear
            const retryButton = await waitFor(() => screen.getByText('Retry'))
            fireEvent.click(retryButton)
            
            expect(defaultProps.onRetry).toHaveBeenCalled()
        })

        it('calls onClose when close button is clicked', async () => {
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={false}
                    generatedImage="test-base64-data"
                />
            )
            
            // Wait for close button to appear
            const closeButton = await waitFor(() => screen.getByText('Close'))
            fireEvent.click(closeButton)
            
            expect(defaultProps.onClose).toHaveBeenCalled()
        })

        it('resets state when retry is clicked', async () => {
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={false}
                    generatedImage="test-base64-data"
                />
            )
            
            // Wait for retry button to appear
            const retryButton = await waitFor(() => screen.getByText('Retry'))
            fireEvent.click(retryButton)
            
            // Should reset to idle state
            await waitFor(() => {
                expect(screen.getByText('Ready to generate')).toBeInTheDocument()
            })
        })
    })

    describe('Generation Sequence', () => {
        it('calls onGenerationStart when generation begins', () => {
            act(() => {
                render(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            })
            
            // Run any pending timers to ensure all effects have run
            act(() => {
                jest.runOnlyPendingTimers()
            })
            
            expect(defaultProps.onGenerationStart).toHaveBeenCalled()
        })

        it('calls onGenerationComplete when generation finishes', async () => {
            const onGenerationComplete = jest.fn()
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={true}
                    onGenerationComplete={onGenerationComplete}
                />
            )
            
            // Fast-forward through the complete generation sequence:
            // 3000ms (progress phase) + 200ms (reveal delay) + 1000ms (complete phase) = 4200ms
            act(() => {
                jest.advanceTimersByTime(4200)
            })
            
            // Run any remaining pending timers to ensure all state updates complete
            act(() => {
                jest.runOnlyPendingTimers()
            })
            
            // Wait for the callback to be called
            await waitFor(() => {
                expect(onGenerationComplete).toHaveBeenCalledTimes(1)
            }, { timeout: 5000 })
        })

        it('handles multiple generation cycles', async () => {
            const onGenerationComplete = jest.fn()
            const onGenerationStart = jest.fn()
            const { rerender } = render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={true}
                    onGenerationComplete={onGenerationComplete}
                    onGenerationStart={onGenerationStart}
                />
            )
            
            // Complete first generation
            act(() => {
                jest.advanceTimersByTime(4200)
            })
            
            // Run any remaining pending timers to ensure all state updates complete
            act(() => {
                jest.runOnlyPendingTimers()
            })
            
            // Wait for the first callback to be called
            await waitFor(() => {
                expect(onGenerationComplete).toHaveBeenCalledTimes(1)
            }, { timeout: 5000 })
            
            // Reset the component state for second generation
            act(() => {
                rerender(
                    <PolaroidPhotoGenerator 
                        {...defaultProps} 
                        isGenerating={false}
                        onGenerationComplete={onGenerationComplete}
                        onGenerationStart={onGenerationStart}
                    />
                )
            })
            
            // Start second generation
            act(() => {
                rerender(
                    <PolaroidPhotoGenerator 
                        {...defaultProps} 
                        isGenerating={true}
                        onGenerationComplete={onGenerationComplete}
                        onGenerationStart={onGenerationStart}
                    />
                )
            })
            
            // Complete second generation
            act(() => {
                jest.advanceTimersByTime(4200)
            })
            
            // Run any remaining pending timers
            act(() => {
                jest.runOnlyPendingTimers()
            })
            
            // Wait for the second callback to be called
            await waitFor(() => {
                expect(onGenerationComplete).toHaveBeenCalledTimes(2)
            }, { timeout: 5000 })
            
            expect(onGenerationStart).toHaveBeenCalledTimes(2)
        })

        it('handles generation with generated image', () => {
            const mockBase64 = 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg=='
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={true}
                    generatedImage={mockBase64}
                />
            )
            
            // Should still call onGenerationStart
            expect(defaultProps.onGenerationStart).toHaveBeenCalled()
        })
    })

    describe('Edge Cases', () => {
        it('works without optional callbacks', () => {
            render(<PolaroidPhotoGenerator isGenerating={true} />)
            
            // Fast-forward to processing phase
            act(() => {
                jest.advanceTimersByTime(100)
            })
            
            // Run any pending timers to ensure state updates
            act(() => {
                jest.runOnlyPendingTimers()
            })
            
            // Should not throw error
            expect(screen.getByText('Processing...')).toBeInTheDocument()
        })

        it('handles empty generated image', () => {
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    generatedImage=""
                />
            )
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('src', 'data:image/png;base64,')
        })

        it('handles null generated image', () => {
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    generatedImage={null as any}
                />
            )
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('src', '/test-mock-image.jpg')
        })

        it('handles undefined generated image', () => {
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    generatedImage={undefined}
                />
            )
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('src', '/test-mock-image.jpg')
        })

        it('handles rapid state changes', () => {
            const { rerender } = render(
                <PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />
            )
            
            // Rapidly change states
            act(() => {
                rerender(<PolaroidPhotoGenerator {...defaultProps} isGenerating={false} />)
            })
            
            act(() => {
                rerender(<PolaroidPhotoGenerator {...defaultProps} isGenerating={true} />)
            })
            
            act(() => {
                rerender(<PolaroidPhotoGenerator {...defaultProps} isGenerating={false} />)
            })
            
            // Run any pending timers to ensure state updates
            act(() => {
                jest.runOnlyPendingTimers()
            })
            
            // Should not throw error and should be in idle state
            expect(screen.getByText('Ready to generate')).toBeInTheDocument()
        })

        it('handles image load error gracefully', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            const img = screen.getByAltText('Generated try-on preview')
            
            // Simulate image load error
            fireEvent.error(img)
            
            // Should not throw error
            expect(img).toBeInTheDocument()
        })
    })

    describe('Accessibility', () => {
        it('has proper ARIA attributes', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} isLoading={true} />)
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toHaveAttribute('aria-busy', 'true')
        })

        it('has proper alt text', () => {
            render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            const img = screen.getByAltText('Generated try-on preview')
            expect(img).toBeInTheDocument()
        })

        it('buttons are accessible', async () => {
            // Test with a component that's already in complete state with generated image
            render(
                <PolaroidPhotoGenerator 
                    {...defaultProps} 
                    isGenerating={false}
                    generatedImage="test-base64-data"
                />
            )
            
            // Wait for buttons to appear
            await waitFor(() => {
                const retryButton = screen.getByRole('button', { name: /retry/i })
                const closeButton = screen.getByRole('button', { name: /close/i })
                
                expect(retryButton).toBeInTheDocument()
                expect(closeButton).toBeInTheDocument()
            })
        })
    })

    describe('Styling and Layout', () => {
        it('has correct polaroid dimensions', () => {
            const { container } = render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            // Use a more specific selector that doesn't rely on Tailwind's arbitrary value syntax
            const polaroid = container.querySelector('div[class*="w-[475px]"]')
            expect(polaroid).toBeInTheDocument()
        })

        it('has correct photo area dimensions', () => {
            const { container } = render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            // Use a more specific selector that doesn't rely on Tailwind's arbitrary value syntax
            const photoArea = container.querySelector('div[class*="h-[400px]"]')
            expect(photoArea).toBeInTheDocument()
        })

        it('has proper shadow styling', () => {
            const { container } = render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            const polaroid = container.querySelector('.shadow-lg')
            expect(polaroid).toBeInTheDocument()
        })

        it('has proper border radius', () => {
            const { container } = render(<PolaroidPhotoGenerator {...defaultProps} />)
            
            const polaroid = container.querySelector('.rounded-sm')
            expect(polaroid).toBeInTheDocument()
        })
    })
}) 