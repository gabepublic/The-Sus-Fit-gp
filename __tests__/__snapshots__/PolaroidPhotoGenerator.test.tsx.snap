// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PolaroidPhotoGenerator Initial Rendering renders placeholder while generatedImage is undefined 1`] = `
<div
  class="relative flex flex-col items-center"
  data-testid="polaroid-generator"
>
  <div
    class="relative bg-white p-6 w-[475px] h-[550px] flex flex-col shadow-lg rounded-sm"
  >
    <div
      class="relative w-full h-[400px] overflow-hidden rounded-sm bg-gray-100"
    >
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <div
          class="text-gray-400 font-mono text-center"
        >
          Ready to generate
        </div>
      </div>
    </div>
    <div
      class="h-[70px] bg-white flex items-center justify-center"
    >
      <div
        class="text-lg text-gray-600 font-mono"
      >
        Ready to generate
      </div>
    </div>
  </div>
</div>
`;
